### PR TITLE
 misskey: init at 2024.5.0 & nixos/misskey: init module

### PR DESCRIPTION
**Pull request by feathecutie**
## Description of changes

This adds the Misskey blogging platform by providing a package and accompanying NixOS module.

Source code: https://github.com/misskey-dev/misskey
Homepage: https://misskey-hub.net/en/

This PR is largely inspired by https://github.com/NixOS/nixpkgs/pull/296697.

### TODO / Nice to have:

- [x] Add the ability to pass secrets via files that are read at runtime.
  - Currently only for `settings.{db.pass,redis.pass,meilisearch.apiKey}`. 
- [x] Correctly type all config options in the NixOS module, as there aren't that many.
  - Created typed options with defaults for the most important options, while leaving the rest freeform. 
- [x] Split out the config options that affect PostgreSQL/Redis/Nginx/Caddy instead of reading them from the untyped `settings` field.
  - Not split out from settings, but all relevant fields are properly typed now.
- [x] Assert the `settings` to be in a state where Misskey doesn't crash at runtime, e.g. when certain options are `null`.
  - All relevant fields have proper types and defaults and can't be null.
- [x] Automatically configure SSL for Nginx (I never used Nginx before).
  - [ ] Potentially detect if SSL is even necessary by checking if `settings.url` starts with `https://` or `http://`. This seems quite error prone, so a seperate toggle for SSL in nginx might make sense. Luckily, Caddy handles this automatically based on the URL type.
  - [ ] Alternative: Remove the option for auto-configuring a Nginx reverse proxy and let the user set it up completely themselves. This seems less ergonomic, but would leave the specifics of setting up SSL (among other things) to the user.

Feedback is welcome for any of the remaining (or even completed) TODOs.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: nixos/modules/services/web-apps/misskey.nix
Comment (feathecutie) [inline nixos/modules/services/web-apps/misskey.nix:259]: This `if` expression breaks the ability to configure meilisearch via `settings.meilisearch`, resulting in `The option `services.misskey.settings.meilisearch` is defined both null and not null`.

The reason why I can't simply use `lib.optionalAttrs` here is that `settings.meilisearch = null` and `settings.meilisearch = { }` are parsed differently by Misskey. `settings.meilisearch` being `null` means Misskey won't try to use meilisearch, while `settings.meilisearch` being `{ }` results in an error.

I'll need to look into how to do this differently.

File: nixos/modules/services/web-apps/misskey.nix
Comment (feathecutie) [inline nixos/modules/services/web-apps/misskey.nix:345]: Setting `reverseProxy.webserver.nginx.enableACME = true` results in the error
```
A definition for option `systemd.services."acme-https://example.com".after."[definition 1-entry 5]"' is not of type `string matching the pattern [a-zA-Z0-9@%:_.\-]+[.](service|socket|device|mount|automount|swap|target|path|timer|scope|slice)
```

I believe this is the case because that the name of the Nginx virtualHost should not include the protocol (`http://` or `https://`). This could be fixed by either removing these prefixes from `cfg.settings.url` when using it as the virtualHost name, or by exposing seperate options for setting the protocol and the domain name itself.

The first option seems very error prone, but the second option does not integrate into the existing configuration system of simply parsing configuration options from `settings`.

I'll test a few different options to try to resolve this and see what feels and works best.

File: pkgs/by-name/mi/misskey/package.nix
Comment (JulienMalka) [inline pkgs/by-name/mi/misskey/package.nix:37]: Could you add a comment here explaining the reason for the patch and if there is a point in time / a condition when it should be dropped ?

**Review by JulienMalka — APPROVED** [review]
Code looks good to me appart from minor comment

**Comment by JulienMalka** [comment]
Before merging please squash your changes so that there is one that adds yourself as maintainer, one that adds the package, one that adds the module

File: pkgs/by-name/mi/misskey/package.nix
Comment (feathecutie) [reply pkgs/by-name/mi/misskey/package.nix]: ..appearently that patch is not necessary after all, I thought I got build failures without it in the beginning but the package seems to build fine without it, so I removed it again.

**Comment by fricklerhandwerk** [comment]
Thanks a lot @feathecutie!

**Pull request merged**
